cmd/kops/create_cluster.go

```diff
@@ -741,3 +793,141 @@ func loadSSHPublicKeys(sshPublicKey string) (map[string][]byte, error) {
+	}
+	tags, err := remote.List(repo)
+	if err != nil {
+		return commandutils.CompletionError("listing kube-apiserver tags", err)
+	}
+	versions := sets.NewString()
+	for _, tag := range tags {
+		parsed, err := kopsutil.ParseKubernetesVersion(tag)
+		if err != nil {
+			continue
+		}
+		if kopsutil.IsKubernetesGTE(cloudup.OldestSupportedKubernetesVersion, *parsed) &&
+			!kopsutil.IsKubernetesGTE(tooNewVersion.String(), *parsed) {
+			versions.Insert(parsed.String())
+		}
+	}
+
+	// Remove pre-release versions that have a subsequent stable version.
+	for _, version := range versions.UnsortedList() {
+		split := strings.Split(version, "-")
+		if len(split) > 1 && versions.Has(split[0]) {
+			versions.Delete(version)
+		}
+	}
+
+	return versions.List(), cobra.ShellCompDirectiveNoFileComp
+}
+
+func completeInstanceImage(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	// TODO call into cloud provider(s) to get list of valid images
+	return nil, cobra.ShellCompDirectiveNoFileComp
+}
+
+func completeMachineType(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	// TODO call into cloud provider(s) to get list of valid machine types
+	return nil, cobra.ShellCompDirectiveNoFileComp
+}
+
+func completeNetworkID(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	// TODO call into cloud provider(s) to get list of valid VPCs
+	return nil, cobra.ShellCompDirectiveNoFileComp
+}
+
+func completeSubnetID(options *CreateClusterOptions) func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		// TODO call into cloud provider(s) to get list of valid Subnet IDs
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+}
+
+func completeStorageType(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	// TODO call into cloud provider(s) to get list of valid storage types
+	return nil, cobra.ShellCompDirectiveNoFileComp
+}
+
+func completeNetworking(options *CreateClusterOptions) func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		completions := []string{
+			"external",
+			"cni",
+			"calico",
+			"cilium",
+			"cilium-etcd",
+		}
+
+		if !options.IPv6 {
+			completions = append(completions,
+				"kubenet",
+				"kopeio",
+				"weave",
+				"flannel",
+				"canal",
+				"kube-router",
+			)
+
+			if options.CloudProvider == "aws" || options.CloudProvider == "" {
+				completions = append(completions, "amazonvpc", "lyftvpc")
+			}
+
+			if cloudup.AlphaAllowGCE.Enabled() && (options.CloudProvider == "gce" || options.CloudProvider == "") {
```

Comment:
I noticed a few odd behaviors with checking the feature flag here.

1. The feature flag needs to be set prior to `kops completion` evaluation rather than at the time of tab completion. I suppose this isn't a big deal since most users would have this particular feature flag set in their default shell environment.
2. With the feature flag set, `kops completion` logs some output that should be silenced:
  ```
  export KOPS_FEATURE_FLAGS=+AlphaAllowGCE
  eval $(./.build/local/kops completion zsh)
  I0718 08:50:04.388774   17741 featureflag.go:167] FeatureFlag "AlphaAllowGCE"=true
  ```

Reply:
I don't think the logging behavior of `kops completion` changed. The old implementation didn't suppress that output either.